src/lib/ccf/index.ts

```diff
@@ -0,0 +1,458 @@
+
+        results.push({
+          e: e,
+          m: m,
+        });
+      });
+    }
+
+    return results;
+  }
+
+  /**
+   * Calculates the energy consumption for a single observation
+   * requires
+   *
+   * duration: duration of the observation in seconds
+   * cpu: cpu usage in percentage
+   * datetime: ISO 8601 datetime string
+   *
+   * Uses a spline method for AWS and linear interpolation for GCP and Azure
+   */
+  private calculateEnergy(observation: KeyValuePair) {
+    if (
+      !('duration' in observation) ||
+      !('cpu' in observation) ||
+      !('datetime' in observation)
+    ) {
+      throw new Error(
+        'Required Parameters duration,cpu,datetime not provided for observation'
+      );
+    }
+
+    //    duration is in seconds
+    const duration = observation['duration'];
+
+    //    convert cpu usage to percentage
+    const cpu = observation['cpu'] * 100.0;
+
+    //  get the wattage for the instance type
+    let wattage;
+
+    if (this.provider === 'aws' && this.interpolation === 'spline') {
+      const x = [0, 10, 50, 100];
+
+      const y: number[] = [
+        this.computeInstances['aws'][this.instanceType].consumption.idle ?? 0,
+        this.computeInstances['aws'][this.instanceType].consumption
+          .tenPercent ?? 0,
+        this.computeInstances['aws'][this.instanceType].consumption
+          .fiftyPercent ?? 0,
+        this.computeInstances['aws'][this.instanceType].consumption
+          .hundredPercent ?? 0,
+      ];
+
+      const spline = new Spline(x, y);
+
+      wattage = spline.at(cpu);
+    } else {
+      const idle =
+        this.computeInstances[this.provider][this.instanceType].consumption
+          .minWatts ?? 0;
+      const max =
+        this.computeInstances[this.provider][this.instanceType].consumption
+          .maxWatts ?? 0;
+
+      // linear interpolation
+      wattage = idle + (max - idle) * (cpu / 100);
+    }
+    //  duration is in seconds
+    //  wattage is in watts
+    //  eg: 30W x 300s = 9000 J
+    //  1 Wh = 3600 J
+    //  9000 J / 3600 = 2.5 Wh
+    //  J / 3600 = Wh
+    //  2.5 Wh / 1000 = 0.0025 kWh
+    //  Wh / 1000 = kWh
+    // (wattage * duration) / (seconds in an hour) / 1000 = kWh
+    return (wattage * duration) / 3600 / 1000;
+  }
+
+  /**
+   * Returns model identifier
+   */
+  modelIdentifier(): string {
+    return 'ccf.cloud.sci';
+  }
+
+  /**
+   * Standardize the instance metrics for all the providers
+   *
+   * Maps the instance metrics to a standard format (min, max, idle, 10%, 50%, 100%) for all the providers
+   */
+  standardizeInstanceMetrics() {
+    this.computeInstances['aws'] = {};
+    this.computeInstances['gcp'] = {};
+    this.computeInstances['azure'] = {};
+    let gcpMin = 0.0;
+    let gcpMax = 0.0;
+    let gcpCount = 0;
+    // standardize gcp emissions
+    // gcp_use loaded from coefficients-gcp-use.csv file from CCF
+    gcp_use.forEach((instance: KeyValuePair) => {
```

Comment:
This function can be simplified with breaking down into small functions, some of which can be reusable
if possible use reduce instead of for each for not declaring `sum variables` on each instance
```
    let gcpMin = 0.0;
    let gcpMax = 0.0;
    let gcpCount = 0;
```